src/libcore/mem/mod.rs

```diff
@@ -932,7 +932,7 @@ pub unsafe fn transmute_copy<T, U>(src: &T) -> U {
 ///
 /// [`discriminant`]: fn.discriminant.html
 #[stable(feature = "discriminant_value", since = "1.21.0")]
-pub struct Discriminant<T>(u64, PhantomData<fn() -> T>);
+pub struct Discriminant<T>(<T as DiscriminantKind>::Discriminant);
```

Comment:
Uh, oh, cc @rust-lang/libs, this changes variance. `PhantomData<fn() -> T>` is covariant wrt `T` but projections are invariant (any "added for future use" parameters should probably start invariant, so we don't break code later).

That means this PR will require a crater run.

Reply:
We could also do the variance change and crater run in a separate PR first.

Reply:
opened #71814